statsd/lexer.go

```diff
@@ -41,35 +60,57 @@ func (l *lexer) next() byte {
 	return b
 }
 
-func (l *lexer) run() (*types.Metric, error) {
+func (l *lexer) run(input []byte, namespace string) (*types.Metric, *types.Event, error) {
+	l.input = input
+	l.namespace = namespace
+	l.len = uint32(len(l.input))
 	l.sampling = float64(1)
 
-	for state := lexKeySep; state != nil; {
+	for state := lexSpecial; state != nil; {
 		state = state(l)
 	}
 	if l.err != nil {
-		return nil, l.err
+		return nil, nil, l.err
 	}
-	if l.m.Type != types.SET {
-		v, err := strconv.ParseFloat(l.m.StringValue, 64)
-		if err != nil {
-			return nil, err
+	if l.m != nil {
+		if l.m.Type != types.SET {
+			v, err := strconv.ParseFloat(l.m.StringValue, 64)
+			if err != nil {
+				return nil, nil, err
+			}
+			if math.IsNaN(v) {
+				return nil, nil, errNaN
+			}
+			l.m.Value = v
+			l.m.StringValue = ""
 		}
-		if math.IsNaN(v) {
-			return nil, errNaN
+		if l.m.Type == types.COUNTER {
+			l.m.Value = l.m.Value / l.sampling
 		}
-		l.m.Value = v
-		l.m.StringValue = ""
+		l.m.Tags = l.tags
+	} else {
+		l.e.Tags = l.tags
 	}
-	if l.m.Type == types.COUNTER {
-		l.m.Value = l.m.Value / l.sampling
-	}
-
-	return l.m, nil
+	return l.m, l.e, nil
 }
 
 type stateFn func(*lexer) stateFn
 
+// check the first byte for special Datadog type.
+func lexSpecial(l *lexer) stateFn {
+	switch b := l.next(); b {
+	case '_':
```

Comment:
Yeah I understand that, I was just unsure if that is done only for the 1st character. If it is not only for the 1st character then it would be a problem with metrics with an _ in the metric name. But I see we have a test for such metric names so I assume it's fine.